File: sdk/python/mockagents/scenario.py
"""Scenario definition and execution for multi-turn conversation testing."""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Optional

from .client import MockAgentClient
from .types import ChatResponse, Interaction


@dataclass
class Scenario:
    """Defines a multi-turn conversation scenario.

    Args:
        name: Scenario name for identification.
        steps: List of message steps to send. Each step is a dict with
            'role' and 'content' keys, plus optional 'tool_calls'.
        agent_name: Target agent name (used for routing).
        protocol: Protocol to use ("openai" or "anthropic").
        model: Model name for the request.
    """

    name: str
    steps: list[dict[str, Any]]
    agent_name: str = ""
    protocol: str = "openai"
    model: str = "gpt-4o"
    session_id: str = field(default_factory=lambda: f"scenario-{uuid.uuid4().hex[:12]}")


@dataclass
class ScenarioResult:
    """Result of executing a scenario.

    Contains all interactions and provides convenience accessors.
    """

    scenario: Scenario
    interactions: list[Interaction] = field(default_factory=list)

    @property
    def responses(self) -> list[ChatResponse]:
        """All response objects from interactions."""
        return [i.response for i in self.interactions]

    @property
    def tool_calls(self) -> list[Any]:
        """Flattened list of all tool calls across all interactions."""
        calls = []
        for interaction in self.interactions:
            calls.extend(interaction.response.tool_calls)
        return calls

    @property
    def latency_ms(self) -> float:
        """Total latency across all interactions."""
        return sum(i.latency_ms for i in self.interactions)

    @property
    def last_response(self) -> Optional[ChatResponse]:
        """The most recent response."""
        if self.interactions:
            return self.interactions[-1].response
        return None

    @property
    def content(self) -> str:
        """Concatenated content from all responses."""
        return " ".join(r.content for r in self.responses if r.content)


def run_scenario(client: MockAgentClient, scenario: Scenario) -> ScenarioResult:
    """Execute a scenario against a MockAgents server.

    Sends each step's user messages sequentially, accumulating the
    conversation history for context.

    Args:
        client: MockAgentClient connected to the server.
        scenario: Scenario to execute.

    Returns:
        ScenarioResult with all interactions.
    """
    result = ScenarioResult(scenario=scenario)
    conversation: list[dict[str, Any]] = []

    for step in scenario.steps:
        conversation.append(step)

        start = time.monotonic()
        if scenario.protocol == "anthropic":
            response = client.message(
                messages=conversation,
                model=scenario.model,
                session_id=scenario.session_id,
            )
        else:
            response = client.chat(
                messages=conversation,
                model=scenario.model,
                session_id=scenario.session_id,
            )
        latency = (time.monotonic() - start) * 1000

        interaction = Interaction(
            request=step,
            response=response,
            latency_ms=latency,
        )
        result.interactions.append(interaction)

        # Add assistant response to conversation for multi-turn context.
        if response.content:
            conversation.append({"role": "assistant", "content": response.content})

    return result

File: sdk/go/mockagents/scenario.go
package mockagents

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"time"
)

// Protocol selects which endpoint a scenario targets.
type Protocol string

const (
	ProtocolOpenAI    Protocol = "openai"
	ProtocolAnthropic Protocol = "anthropic"
)

// ScenarioStep is one message in a scripted conversation.
type ScenarioStep struct {
	Role    string
	Content string
}

// Scenario is a declarative multi-turn test case.
type Scenario struct {
	Name      string
	Steps     []ScenarioStep
	Protocol  Protocol
	SessionID string
	Model     string
}

// NewScenario constructs a Scenario, defaulting to OpenAI protocol and
// a random session id.
func NewScenario(name string, steps []ScenarioStep) *Scenario {
	if name == "" {
		panic("mockagents: scenario name is required")
	}
	if len(steps) == 0 {
		panic("mockagents: scenario must have at least one step")
	}
	return &Scenario{
		Name:      name,
		Steps:     steps,
		Protocol:  ProtocolOpenAI,
		SessionID: randomSessionID(),
	}
}

// ScenarioResult aggregates every response produced by RunScenario.
type ScenarioResult struct {
	ScenarioName   string
	Responses      []*ChatResponse
	TotalLatencyMs float64
}

// Last returns the final response, or nil if the scenario produced none.
func (r *ScenarioResult) Last() *ChatResponse {
	if len(r.Responses) == 0 {
		return nil
	}
	return r.Responses[len(r.Responses)-1]
}

// LastContent is a convenience for pulling content off the final response.
func (r *ScenarioResult) LastContent() string {
	last := r.Last()
	if last == nil {
		return ""
	}
	return last.Content
}

// RunScenario walks the scenario steps against the given client. Each
// user step triggers a single request; assistant/system steps are
// recorded as prior context only. Assistant replies from the server are
// folded back into the conversation history so multi-turn scenarios see
// the same state the server does.
func RunScenario(ctx context.Context, client *Client, scenario *Scenario) (*ScenarioResult, error) {
	if client == nil {
		return nil, errors.New("mockagents: client is nil")
	}
	if scenario == nil {
		return nil, errors.New("mockagents: scenario is nil")
	}

	history := make([]ChatMessage, 0, len(scenario.Steps)*2)
	result := &ScenarioResult{ScenarioName: scenario.Name}
	start := time.Now()

	for _, step := range scenario.Steps {
		history = append(history, ChatMessage{Role: step.Role, Content: step.Content})
		if step.Role != "user" {
			continue
		}
		var (
			resp *ChatResponse
			err  error
		)
		switch scenario.Protocol {
		case ProtocolAnthropic:
			resp, err = client.Message(ctx, append([]ChatMessage(nil), history...), MessageOptions{
				Model:     scenario.Model,
				SessionID: scenario.SessionID,
			})
		default:
			resp, err = client.Chat(ctx, append([]ChatMessage(nil), history...), ChatOptions{
				Model:     scenario.Model,
				SessionID: scenario.SessionID,
			})
		}
		if err != nil {
			return nil, fmt.Errorf("step %d (%q): %w", len(result.Responses)+1, step.Content, err)
		}
		result.Responses = append(result.Responses, resp)
		history = append(history, ChatMessage{Role: "assistant", Content: resp.Content})
	}

	result.TotalLatencyMs = float64(time.Since(start).Microseconds()) / 1000.0
	return result, nil
}

func randomSessionID() string {
	var b [8]byte
	_, _ = rand.Read(b[:])
	return "scenario-" + hex.EncodeToString(b[:])
}
